windows_service/datadog_checks/windows_service/windows_service.py

```diff
@@ -126,6 +165,36 @@ def is_delayed_auto(self):
             )
         return self._is_delayed_auto
 
+    @property
+    def trigger_count(self):
+        if self._trigger_count is None:
+            # find out how many bytes to allocate for buffer
+            # raise error if the error code is not ERROR_INSUFFICIENT_BUFFER
+            bytesneeded = ctypes.c_uint32(0)
+            try:
+                QueryServiceConfig2W(
+                    ctypes.c_void_p(self.hSvc.handle), SERVICE_CONFIG_TRIGGER_INFO, None, 0, ctypes.byref(bytesneeded)
+                )
+            except OSError as e:
+                if e.winerror != winerror.ERROR_INSUFFICIENT_BUFFER:
+                    raise
+
+            # allocate buffer and get trigger info, raise error if function returns 0
+            bytesBuffer = ctypes.create_string_buffer(bytesneeded.value)
+            QueryServiceConfig2W(
```

Comment:
The call to QueryServiceConfig2W could fail at this point and we should handle this case before moving forward with the triggerStruct assignment on line 193?

Reply:
Hi @lrbhatia , When `QueryServiceConfig2W` fails it will raise an OSError. Since any OSError thrown at this point should just be raised we didn't require a try/except. The OSError thrown at line 184 would be caught on line 264. 

In `test_windows_service.py` the function `test_trigger_count_failure` tests for the OSError to be thrown upon failure.

Reply:
I have made the same "error" of judgement as @lrbhatia and added a comment and then realized the fact you are describing and delete the review comment. The appearance of asymmetry of handling the error is thrown unsuspected reviewer of. I can suggest adding a comment about it but better yet IMHO move that specific error handling and rethrowing completely into our QueryServiceConfig2W. BTW, did you have a chance of simulating the failure and throwing the error, how it is handled above and do we log it?

Reply:
The following lines starting at 264 show how the errors are handled:
```
                      except (pywintypes.error, OSError) as e:
                        self.log.exception("Exception at service match for %s", service_filter)
                        self.warning("Failed to query %s service config for filter %s: %s", short_name, service_filter, str(e))
```
Basically when running the `service_filter.match`, this will cause the integration to query for the necessary service info to find a match. If any of the queries raise a pywintypes.error or OSError it will be caught and added as a warning log while the integration continues to run. 

I simulated the error handling in `test_trigger_count_failure` with:
```
patch(
        'datadog_checks.windows_service.windows_service.QueryServiceConfig2W',
        side_effect=[ctypes.WinError(winerror.ERROR_INSUFFICIENT_BUFFER), ctypes.WinError(1)] * 2,
    )
```
With these side effects, there should be a `ctypes.WinError(1)` exception raised which is searched for in caplog.text. The side effects were multiplied by 2 to account for the fact that `QueryServiceConfig2W` is run twice each to query for two services.

Reply:
That is awesome 👍

what about first part of suggestion
> The appearance of asymmetry of handling the error is thrown unsuspected reviewer of. I can suggest adding a comment about it but better yet IMHO move that specific error handling and rethrowing completely into our QueryServiceConfig2W.

Reply:
I can definitely leave a comment specifying why there isn't a try/except. 

Could you elaborate on how to have error handling be done in `QueryServiceConfig2W`? Would it be something like having both instances of `ctypes.windll.advapi32.QueryServiceConfig2W(*args)` be run inside of `QueryServiceConfig2W` and either raise an error or return the triggerCount?

Reply:
Ah, right, sorry I was not too specific. Your above suggestion is just as good or better (albeit bit more work) but what I meant to have something like this which makes minimal changes to your existing code/flow IMHO (I did not compile it)

```
def QueryServiceConfig2W(*args):
    """
    ctypes wrapper for info types not supported by pywin32
    """
    if ctypes.windll.advapi32.QueryServiceConfig2W(*args) == 0:
        # raise error if function returns 0 unless we are calling to get requiired buffer size and error is expected ERROR_INSUFFICIENT_BUFFER
        if *args[2] != None || ctypes.WinError() != winerror.ERROR_INSUFFICIENT_BUFFER:
            raise ctypes.WinError()

```
.... and accordingly, without try/catch.
```
    @property
    def trigger_count(self):
        if self._trigger_count is None:
            # find out how many bytes to allocate for buffer
            bytesneeded = ctypes.c_uint32(0)
            QueryServiceConfig2W(
                    ctypes.c_void_p(self.hSvc.handle), SERVICE_CONFIG_TRIGGER_INFO, None, 0, ctypes.byref(bytesneeded)
            )

            # allocate buffer and get trigger info
            bytesBuffer = ctypes.create_string_buffer(bytesneeded.value)
            QueryServiceConfig2W(
                ctypes.c_void_p(self.hSvc.handle),
                SERVICE_CONFIG_TRIGGER_INFO,
                ctypes.byref(bytesBuffer),
                bytesneeded,
                ctypes.byref(bytesneeded),
            )

            # converting returned buffer into TriggerInfo to get trigger count
            triggerStruct = TriggerInfo.from_buffer(bytesBuffer)
            self._trigger_count = triggerStruct.triggerCount

        return self._trigger_count
```

Reply:
On the related note, your comment IMHO, and I did not notice it before, 
```
            # allocate buffer and get trigger info, raise error if function returns 0
            bytesBuffer = ctypes.create_string_buffer(bytesneeded.value)
            QueryServiceConfig2W(

```
is a bit misleading or at least could be ambiguous everything considered, don't you agree?

Reply:
>  ```        
>  if *args[2] != None || ctypes.WinError() != winerror.ERROR_INSUFFICIENT_BUFFER:
>             raise ctypes.WinError()
>  ```

This would make the second call succeed if it returned `ERROR_INSUFFICIENT_BUFFER`, even though we would want it to fail.

Reply:
Correct, I did not think it through, perhaps below is easier to "read" and correct
```
def QueryServiceConfig2W(*args):
    """
    ctypes wrapper for info types not supported by pywin32
    """
    rc = ctypes.windll.advapi32.QueryServiceConfig2W(*args) 
    if *args[2] == None:
        # raise error if getting buffer size calculation call pattern did not return expected results
        if !(rc == 0 && ctypes.WinError() == winerror.ERROR_INSUFFICIENT_BUFFER):
            raise ctypes.WinError()
    else:
        # raise error if the call to get service config data into provide sufficient size bugger fails
        if rc == 0:
            raise ctypes.WinError()
```

Reply:
Please merge now. It is a banal improvement suggestion and we may consider it in future. Thank you @mrafi97

Reply:
Hey @iglendd apologies for the delay. I didn't get a proper chance to work on this PR in the past week. I just added an updated comment for the time being.

Reply:
Do not worry. We figure that it was minor (and subjective) improvements and hence you are not embedded anymore we did not want to keep engaging you on that. It is ready to merge.